include/openssl/hpke.h

```diff
@@ -0,0 +1,499 @@
+ */
+int OSSL_HPKE_CTX_set1_ikme(OSSL_HPKE_CTX *ctx,
+                            const unsigned char *ikme, size_t ikmelen);
+
+/**
+ * @brief set a sender private key for HPKE authenticated modes
+ * @param ctx is the pointer for the HPKE context
+ * @param privp is an EVP_PKEY form of the private key
+ * @return 1 for success, 0 for error
+ */
+int OSSL_HPKE_CTX_set1_authpriv(OSSL_HPKE_CTX *ctx, EVP_PKEY *privp);
+
+/**
+ * @brief set a public key for HPKE authenticated modes
+ * @param ctx is the pointer for the HPKE context
+ * @param pub is an buffer form of the public key
+ * @param publen is the length of the above
+ * @return 1 for success, 0 for error
+ *
+ * In all these APIs public keys are passed as buffers whereas
+ * private keys as passed as EVP_PKEY pointers.
+ */
+int OSSL_HPKE_CTX_set1_authpub(OSSL_HPKE_CTX *ctx,
+                               const unsigned char *pub,
+                               size_t publen);
+
+/**
+ * @brief ask for the state of the sequence of seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq returns the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The value returned is the next one to be used when sealing
+ * or opening (so as we start at zero this will be 1 after the
+ * first successful call to seal or open)
+ *
+ * seq is a uint64_t as that's what two other implementations
+ * chose
+ */
+int OSSL_HPKE_CTX_get0_seq(OSSL_HPKE_CTX *ctx, uint64_t *seq);
+
+/**
+ * @brief set the sequence value for seal/open calls
+ * @param ctx is the pointer for the HPKE context
+ * @param seq set the positive integer sequence number
+ * @return 1 for success, 0 for error
+ *
+ * The next seal or open operation will use this value.
+ */
+int OSSL_HPKE_CTX_set1_seq(OSSL_HPKE_CTX *ctx, uint64_t seq);
+
+/**
+ * @brief sender seal function
+ * @param ctx is the pointer for the HPKE context
+ * @param enc is the sender's ephemeral public value
+ * @param enclen is the size the above
+ * @param ct is the ciphertext output
+ * @param ctlen is the size the above
+ * @param pub is the recipient public key octets
+ * @param publen is the size the above
+ * @param info is the info parameter
+ * @param infolen is the size the above
+ * @param aad is the aad parameter
+ * @param aadlen is the size the above
+ * @param pt is the plaintext
+ * @param ptlen is the size the above
+ * @return 1 for success, 0 for error
+ *
+ * This can be called once, or multiple, times.
+ *
+ * If no KEM private key has been set in the context an ephemeral
+ * key will be generated and used for the duration of the context.
+ *
+ * The ciphertext buffer (ct) should be big enough to include
+ * the AEAD tag generated from encryptions and the ``enc`` buffer
+ * (the ephemeral public key) needs to be big enough for the
+ * relevant KEM. ``OSSL_HPKE_expansion`` can be used to determine
+ * the sizes needed.
+ */
+int OSSL_HPKE_sender_seal(OSSL_HPKE_CTX *ctx,
```

Comment:
I am wondering if we should just keep this simple and remove OSSL_HPKE_sender_seal() and OSSL_HPKE_recipient_open().. The pod file for HPKE could then just have an example section showing the one usage.

Reply:
yep, as I said, that'll be done in next push